src/setup/testnet/mod.rs

```diff
@@ -0,0 +1,199 @@
+//! Utilities for setting up a testnet consisting of 3 nodes.
+
+use std::{
+    fmt,
+    fmt::Write,
+    fs, io,
+    net::{IpAddr, SocketAddr},
+    path::{Path, PathBuf},
+};
+
+use crate::setup::{
+    config::{CONFIG_FILE, DEFAULT_PORT, ZIGGURAT_CONFIG},
+    node::Node,
+};
+
+/// Testnet's directory for nodes' configs.
+const TESTNET_DIR: &str = "testnet";
+
+/// Validators file name.
+const VALIDATORS_FILE_NAME: &str = "validators.txt";
+
+/// A struct allowing for convenient starting and stopping a small testnet.
+pub struct TestNet {
+    // Setup information for each node. Used for writing configuration.
+    setups: [NodeSetup; 3],
+    // Running nodes. Used to stopping the testnet.
+    running: Vec<Node>,
+    // Sets whether to log the node's output to Ziggurat's output stream.
+    use_stdout: bool,
+}
+
+impl Default for TestNet {
+    fn default() -> Self {
+        Self {
+            setups: [
+                NodeSetup::new(
+                    "127.0.0.1".parse().unwrap(),
+                    "nHUSqn9qjEF7JJkVqvY7BFLMKdqP5KLLEjo5oB4QH43ADDndRawB".into(),
+                    include_str!("token1.txt").into(),
+                ),
+                NodeSetup::new(
+                    "127.0.0.2".parse().unwrap(),
+                    "nHUEsvSFTf1Snr7ZUdLxjcMW6PKcMrwwXCGZBg6xb1ePG8R4C3TS".into(),
+                    include_str!("token2.txt").into(),
+                ),
+                NodeSetup::new(
+                    "127.0.0.3".parse().unwrap(),
+                    "nHUuYdS49cPfRmCXPTwu7MVVFZFFmfG7y5sRttirVMhwuD7xStQp".into(),
+                    include_str!("token3.txt").into(),
+                ),
+            ],
+            running: vec![],
+            use_stdout: false,
+        }
+    }
+}
+
+impl TestNet {
+    /// Starts a testnet.
+    pub async fn start(&mut self) -> anyhow::Result<()> {
+        self.cleanup().await?;
+        let validators_contents = self.build_validators_file_contents().await?;
+        for (i, setup) in self.setups.iter().enumerate() {
+            let node = self
+                .start_node(&(i + 1).to_string(), setup, &validators_contents)
+                .await?;
+            self.running.push(node);
+        }
+        Ok(())
+    }
+
+    /// Stops the testnet.
+    pub async fn stop(mut self) -> anyhow::Result<()> {
+        self.running.iter_mut().for_each(|node| {
+            if let Err(e) = node.stop() {
+                eprintln!("Unable to stop node: {:?}", e);
+            }
+        });
+        self.cleanup().await?;
+        Ok(())
+    }
+
+    // Creates `validators.txt` file with keys of all nodes.
+    async fn build_validators_file_contents(&self) -> Result<String, fmt::Error> {
```

Comment:
```suggestion
    async fn create_validators_file(&self) -> Result<String, fmt::Error> {
```

Reply:
The thing is, it doesn't really create the file, that's done in `write_validators_file`. The reason for the split is that all three nodes take the same `validators.txt` file, but it's easier to have 3 files for one in each node subdir (due to how paths are treated inside Node).